// @kintone/customize-uploader/dist/params/init.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.inquireInitParams = void 0;
const inquirer = __importStar(require("inquirer"));
const messages_1 = require("../messages");
const inquireInitParams = (lang) => {
    const appId = "";
    const scope = "";
    const m = (0, messages_1.getBoundMessage)(lang);
    const questions = [
        {
            type: "input",
            message: m("Q_AppId"),
            name: "appId",
            validate: (v) => !!v,
        },
        {
            type: "list",
            message: m("Q_Scope"),
            name: "scope",
            choices: ["ALL", "ADMIN", "NONE"],
        },
    ];
    return inquirer
        .prompt(questions)
        .then((answers) => Object.assign({ appId, scope, lang }, answers));
};
exports.inquireInitParams = inquireInitParams;
