// @kintone/rest-api-client/esm/src/client/SpaceClient.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { BaseClient } from "./BaseClient";
var SpaceClient = /** @class */ (function (_super) {
    __extends(SpaceClient, _super);
    function SpaceClient() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SpaceClient.prototype.getSpace = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space",
        });
        return this.client.get(path, params);
    };
    SpaceClient.prototype.deleteSpace = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space",
        });
        return this.client.delete(path, params);
    };
    SpaceClient.prototype.updateSpaceBody = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space/body",
        });
        return this.client.put(path, params);
    };
    SpaceClient.prototype.getSpaceMembers = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space/members",
        });
        return this.client.get(path, params);
    };
    SpaceClient.prototype.updateSpaceMembers = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space/members",
        });
        return this.client.put(path, params);
    };
    SpaceClient.prototype.updateThread = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space/thread",
        });
        return this.client.put(path, params);
    };
    SpaceClient.prototype.addThreadComment = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space/thread/comment",
        });
        return this.client.post(path, params);
    };
    SpaceClient.prototype.addGuests = function (params) {
        var path = this.buildPath({
            endpointName: "guests",
        });
        return this.client.post(path, params);
    };
    SpaceClient.prototype.deleteGuests = function (params) {
        var path = this.buildPath({
            endpointName: "guests",
        });
        return this.client.delete(path, params);
    };
    SpaceClient.prototype.updateSpaceGuests = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "space/guests",
        });
        return this.client.put(path, params);
    };
    SpaceClient.prototype.addSpaceFromTemplate = function (params) {
        var path = this.buildPath({
            endpointName: "template/space",
        });
        return this.client.post(path, params);
    };
    return SpaceClient;
}(BaseClient));
export { SpaceClient };
